### ISSUE TITLE
[Spot/Serve] Optimize the translation of filemounts

### ISSUE THREAD

**Issue by Michaelvll** [skypilot-org/skypilot#3188]
The filemounts can contain large amount of small files in the source file. We should consider compression those files while translating them to bucket uploads to optimize the upload/download speed. Since this whole process is hidden to the user, it should be fine to have the intermediate bucket storing the compressed file from the source.

https://github.com/skypilot-org/skypilot/blob/bc75e885a199cb815a18078164d0fda25a25979f/sky/utils/controller_utils.py#L384

**Comment by warrickhe** [comment]
Why don't we compress all the files being sent across? Wouldn't this save even more cost whilst also remaining as a simpler change? We could also implement this as a toggle (in the case that there isn't enough disk space for the user).

**Comment by cblmemo** [comment]
> Why don't we compress all the files being sent across? Wouldn't this save even more cost whilst also remaining as a simpler change? We could also implement this as a toggle (in the case that there isn't enough disk space for the user).

Yeah, compressing all files being sent is a good idea. We should still decompress the file on the cloud VM though, so the user can use file mounts as usual.

Related to disk space cannot accommodating user files, it would be great to do some pre-checks and calculate its total disk space usage, and compare it to the configured disk size. If the provisioned disk space is not enough we should raise an error before any provisioning happens.

**Comment by warrickhe** [comment]
I just found out that rsync has a compression option, -z. Is there any particular reason why we are not using it currently?

**Comment by cblmemo** [comment]
> I just found out that rsync has a compression option, -z. Is there any particular reason why we are not using it currently?

Could you elaborate on this option and do some tests? e.g. can it actually optimize the time to upload a large amount of small files

**Comment by Michaelvll** [comment]
The main issue with not using rsync is the nature of jobs/serve. During recovery, we have to download the files/folders somewhere, and with `rsync` we can only do so when the client is always there, which is not true.
